File: LLM/LLM/Core/FeedForward.cs
using TorchSharp;
using static TorchSharp.torch;

namespace LLM.Core;

public class FeedForward : nn.Module
{
    private readonly nn.Sequential net;

    // Dropout pode ser adicionado como parâmetro
    public FeedForward(string name, int embedDim, double dropout = 0.1) : base(name)
    {
        net = nn.Sequential(
            ("linear1", nn.Linear(embedDim, 4 * embedDim)),
            ("activation", nn.GELU()), // GELU é comum em Transformers modernos
            ("linear2", nn.Linear(4 * embedDim, embedDim)),
            ("dropout", nn.Dropout(dropout)) // Dropout após a segunda linear é comum
        );
        RegisterComponents();
    }

    public override Tensor forward(Tensor x) => net.forward(x);
}
// --- END OF FILE FeedForward.cs ---

File: AI/LLM/core/feed_forward.py
# core/feed_forward.py

import torch
import torch.nn as nn # Importar nn

class FeedForward(nn.Module):
    """
    Rede Feed-Forward usada em um bloco Transformer.
    """
    def __init__(self, embed_dim: int, dropout: float = 0.1):
        super().__init__() # Primeira linha!

        self.net = nn.Sequential(
            nn.Linear(embed_dim, 4 * embed_dim),
            nn.GELU(),
            nn.Linear(4 * embed_dim, embed_dim),
            nn.Dropout(dropout)
        )

    def forward(self, x: torch.Tensor) -> torch.Tensor:
        ff_output = self.net(x)
        return ff_output # Retornar tensor